x/stake/keeper/validator.go

```diff
@@ -120,6 +120,16 @@ func (k Keeper) SetNewValidatorByPowerIndex(ctx sdk.Context, validator types.Val
 	store.Set(GetValidatorsByPowerIndexKey(validator), validator.OperatorAddr)
 }
 
+// Set validator bond height/intratxcounter.  In case of a conflict, the
+// validator which least recently changed power takes precedence.
+// NOTE: Mutates the validator.
+func (k Keeper) BumpValidatorBondHeightAndCounter(ctx sdk.Context, validator *types.Validator) {
```

Comment:
Passing in pointers and modifying those objects is the type of programming pattern which I'd like to avoid unless there is an important performance consideration. In this situation, it's not so bad and pretty straightforward, however I've seen some seriously confusing spaghetti code which we've produced in the past using this pattern. So yeah I've got a reasonably strong preference for always passing back modified inputs instead of silently modifying them via pointer inputs.

Reply:
It's only because we aren't already using pointers for validators... We should discuss this in the future when we have more time, but passing non-pointer structs has worse issues in terms of code readability and analysis of correctness... and it's semantically uglier as well.  I'd rather we use pointers more, but something to discuss in the future.